packages/salesforce-adapter/src/filters/foreign_key_references.ts

```diff
@@ -0,0 +1,126 @@
+ */
+const convertAnnotationsToReferences = (
+  elements: Element[],
+  typeToElemID: Record<string, ElemID>,
+): void => {
+  const resolveTypeReference = (ref: string | ReferenceExpression):
+    string | ReferenceExpression => {
+    if (_.isString(ref)) {
+      const referenceElemId = typeToElemID[ref]
+      if (referenceElemId !== undefined) {
+        return new ReferenceExpression(referenceElemId)
+      }
+    }
+    return ref
+  }
+
+  elements
+    .filter(isObjectType)
+    .flatMap((obj: ObjectType) => Object.values(obj.fields))
+    .filter((field: Field) => field.annotations[FOREIGN_KEY_DOMAIN] !== undefined)
+    .forEach((field: Field): void => {
+      field.annotations[FOREIGN_KEY_DOMAIN] = field.annotations[FOREIGN_KEY_DOMAIN]
+        .map(resolveTypeReference)
+    })
+}
+
+/**
+ * Resolve references using the mapping genrated from the Salesforce DescribeValueType API.
+ *
+ * @param instance                The current instance being modified
+ * @param apiNameToElemIDs        Known element ids, mapped by API name and metadata type
+ */
+const resolveReferences = (
+  instance: InstanceElement,
+  apiNameToElemIDs: ApiNameMapping,
+): void => {
+  const shouldReplace = (field: Field): boolean => (
+    makeArray(field.annotations[FOREIGN_KEY_DOMAIN]).some(isReferenceExpression)
+  )
+
+  const replacePrimitive = (val: Value, field: Field): Value => {
+    if (val === undefined || !_.isString(val)) {
+      return val
+    }
+
+    const refTarget = field.annotations[FOREIGN_KEY_DOMAIN]
+      .filter(isReferenceExpression)
+      .map((ref: ReferenceExpression) => apiNameToElemIDs[val]?.[
+        ref.elemId.typeName
+      ])
+      .find(values.isDefined)
+    return refTarget !== undefined ? new ReferenceExpression(refTarget) : val
+  }
+
+  const transformPrimitive: TransformFunc = ({ value, field }) => (
+    field !== undefined && shouldReplace(field) ? replacePrimitive(value, field) : value
+  )
+
+  // not using transformElement because we're editing the instance in-place
+  instance.value = transformValues({
+    values: instance.value,
+    type: instance.type,
+    transformFunc: transformPrimitive,
+    strict: false,
+  }) ?? instance.value
+}
+
+export const apiNameToElemID = (elements: Element[]): Record<string, ElemID> => (
+  _(elements)
+    .filter(isObjectType)
+    .map(e => [apiName(e), e.elemID])
+    .fromPairs()
+    .value()
+)
+
+/**
+ * Use annotations generated from the DescribeValueType API foreignKeyDomain data to resolve
+ * names into reference expressions.
+ */
+const filter: FilterCreator = () => ({
+  onFetch: async (elements: Element[]) => {
+    const typeToElemID = apiNameToElemID(elements)
+    convertAnnotationsToReferences(elements, typeToElemID)
+
+    const apiNameToElemIDs = groupByAPIName(elements)
+    elements.filter(isInstanceElement).forEach((instance: InstanceElement) => {
+      resolveReferences(instance, apiNameToElemIDs)
+    })
```

Comment:
can let typescript infer the types
```suggestion
    elements
      .filter(isInstanceElement)
      .forEach(instance => {
        resolveReferences(instance, apiNameToElemIDs)
      })
```